### DISCUSSION TITLE
Mongoose Websocket Server with JavaScript Websocket Client?

### DISCUSSION THREAD

**Discussion by efweber999** [cesanta/mongoose#1685] [Ask questions, get answers] [answered]
Hi,

I have an application where I would like to use a Mongoose websocket server, but it needs to communicate with an existing JavaScript websocket client. I see that the way in which Mongoose upgrades to websocket is based on having "/websocket" in the uri. So the example websocket server does not connect to the existing JavaScript websocket client. I'm sure I can get around this, but my question before investing time is; is the rest of the websocket server code compliant with the websocket standard, or will communicating with the JavaScript client have additional issues?

Here is a simple version of the JavaScript websocket client. Many Thanks.

Index.html
--------------
```
<!DOCTYPE html>

<html lang="en">
 <head>
  <meta charset="utf-8" />

  <title>Simple Websocket Client</title>
  <meta name="description" content="Basic template" />
 </head>

 <body>
  <script src="/index.js"></script>
 </body>
</html>
```

index.js
----------
```
alert("[message] Will now upgrade to websocket.");

var socket = new WebSocket('ws://localhost:9000');

socket.onopen = function(e) {
  alert("[open] Connection established");
  alert("Sending message to server");
  socket.send("The quick brown fox jumps over the lazy dog");
};

socket.onmessage = function(event) {
  alert(`[message] Data received from server: ${event.data}`);
};

socket.onclose = function(event) {
  if (event.wasClean) {
    alert(`[close] Connection closed cleanly, code=${event.code} reason=${event.reason}`);
  } else {
    alert('[close] Connection died');
  }
};

socket.onerror = function(error) {
  alert(`[error] ${error.message}`);
};
```

**Comment by scaprile** [comment, marked as answer]
[RFC6455 section 3](https://www.rfc-editor.org/rfc/rfc6455#section-3) clearly allows URI paths.
Observe the example code, you control your URIs, you decide where to upgrade. The choice of using `/websocket` is just a convenience to be able to also serve HTML using a single port (and a single listener). How do you plan to serve that client to the browser ?
If you don't want to serve HTML then just upgrade on any URI, just change https://github.com/cesanta/mongoose/blob/7252f5385b35ce16321f6920d9f2944550a5bded/examples/websocket-server/main.c#L18-L32 to
```
} else if (ev == MG_EV_HTTP_MSG) {
    struct mg_http_message *hm = (struct mg_http_message *) ev_data;
    mg_ws_upgrade(c, hm, NULL);
} else if (ev == MG_EV_WS_MSG) {
```
You might as well modify your JS client by adding `/websocket` after `:9000`
Otherwise, add another listener at port 9000 with its own event handler as the one above; you'll serve HTTP on port 8000 and WS on port 9000

**Reply by efweber999** [reply]
Thanks for the quick response!

Thanks also for clarification on the spec. I'm learning this, and I think I'm assuming it's more complicated than it actually is. Actually, I've looked at multiple websocket libraries and I think everyone makes it more complicated than Mongoose does.

I need to serve the HTML/Javascript to the client and then upgrade to websocket. So I'm going to add "/websocket" to ws://localhost:9000, and go from there.

Best

**Reply by efweber999** [reply]
Using the stock websocket-server.c
That was easy. ;)
![Screenshot from 2022-08-26 11-47-32](https://user-images.githubusercontent.com/81587911/186944187-6eb8f384-bba5-4937-bd53-5fcd701cad99.png)

**Reply by scaprile** [reply]
Mongoose is designed for ease of use
